File: Human_written_code/Code_with_comments/CPP/Q37.cpp
#include <bits/stdc++.h>
using namespace std;

// Function to check if a string has valid balanced parentheses
bool isValid(string s) {
    stack<char> st;

    // Iterate through each character in the string
    for (auto it : s) {
        // Push opening brackets onto the stack
        if (it == '(' || it == '{' || it == '[') 
            st.push(it);
        else {
            // If stack is empty and a closing bracket is encountered, it's invalid
            if (st.empty()) return false;

            char ch = st.top();
            st.pop();

            // Check if the closing bracket matches the top opening bracket
            if ((it == ')' && ch == '(') || (it == ']' && ch == '[') || (it == '}' && ch == '{'))
                continue;
            else
                return false;
        }
    }

    // If the stack is empty, all brackets are matched; otherwise, it's invalid
    return st.empty();
}

int main() {
    string s = "()[{}()]";

    // Output result based on the validity of the string
    if (isValid(s))
        cout << "True" << endl;
    else
        cout << "False" << endl;
}

File: Human_written_code/Code_with_comments/Python/Q37.py
def isValid(s: str) -> bool:
    st = []  # Stack to store opening brackets
    for it in s:
        # Push opening brackets onto the stack
        if it == '(' or it == '{' or it == '[':
            st.append(it)
        else:
            # If stack is empty and a closing bracket is encountered, it's invalid
            if len(st) == 0:
                return False
            ch = st[-1]  # Check the top element of the stack
            st.pop()  # Remove the top element
            # Check if the closing bracket matches the top opening bracket
            if (it == ')' and ch == '(') or (it == ']' and ch == '[') or (it == '}' and ch == '{'):
                continue
            else:
                return False
    # If the stack is empty, all brackets are matched; otherwise, it's invalid
    return len(st) == 0

if __name__ == '__main__':
    s = "()[{}()]"
    # Output result based on the validity of the string
    if isValid(s):
        print("True")
    else:
        print("False")
